src/comlink.ts

```diff
@@ -148,22 +155,54 @@ export function expose(obj: any, ep: Endpoint = self as any) {
     }
     const [wireValue, transferables] = toWireValue(returnValue);
     ep.postMessage({ ...wireValue, id }, transferables);
-  }) as any);
+    if (type === MessageType.RELEASE) {
+      // detach and deactive after sending release response above.
+      ep.removeEventListener("message", callback as any);
+      deactiveEndPoint(ep);
+    }
+  } as any);
   if (ep.start) {
     ep.start();
   }
 }
 
+function isMessagePort(endpoint: Endpoint): endpoint is MessagePort {
+  return endpoint.constructor.name === "MessagePort";
+}
+
+function deactiveEndPoint(endpoint: Endpoint) {
+  if (isMessagePort(endpoint)) endpoint.close();
+}
+
 export function wrap<T>(ep: Endpoint): Remote<T> {
   return createProxy<T>(ep) as any;
 }
 
+function throwIfProxyReleased(isReleased: boolean) {
+  if (isReleased) {
+    throw new Error("Proxy has been released and is not useable");
+  }
+}
+
 function createProxy<T>(
   ep: Endpoint,
   path: (string | number | symbol)[] = []
 ): Remote<T> {
-  const proxy: Function = new Proxy(function() {}, {
+  let isProxyReleased = false;
+  const proxy = new Proxy(function() {}, {
     get(_target, prop) {
+      throwIfProxyReleased(isProxyReleased);
```

Comment:
What if we kept a `WeakMap` of released endpoints and we just do a check in `requestResponseMessage()` instead?

Reply:
@surma  can u elaborate here - WeakMap seems a bit more complicated rather than just relying on simple closure var 'isProxyReleased' - keeps everything tightly controlled in Proxy.  but i am open to suggestions here - maybe misunderstood your suggestion.

Reply:
Hm, I was trying to make the code smaller, but really gzip will take care of that well enough, so disregard this.